File: src/main/core/itunessd.ts
/**
 * iTunesSD 解析与序列化 —— iPod Shuffle 的播放数据库。
 *
 * 重要：shuffle 播放读的是 `iPod_Control/iTunes/iTunesSD`（根 chunk `bdhs`），
 * **不是** `iTunesDB`。`iTunesDB` 只是 Apple 自己的记账文件，本工具只读它取标题。
 *
 * 该格式无需任何签名 / 加密哈希（经典 iPod 的 hash58/hash72/hashAB 在 shuffle 上不存在），
 * 这是纯软件实现可行的根本原因。
 *
 * 本实现是 `tools/itunessd-reference.py` 的 TypeScript 移植，
 * 已通过「解析设备原文件 → 重新序列化 → 逐字节比对」验证（回归测试见 scripts/test-core.js）。
 *
 * 文件布局
 * --------
 *   [0x00]                    根头 64 字节（bdhs）
 *   [0x40]                    hths 曲目头：20 字节 + n×4 偏移表
 *   [0x40 + 20 + n*4]         n × 372 字节曲目记录（rths）
 *   [playlistHeaderOffset]    hphs 播放列表头：68 + m×4 字节
 *   [lphs...]                 m 条播放列表记录（lphs）
 */
import {
  FileType,
  PlaylistRecord,
  SdModel,
  SdRoot,
  TrackRecord,
} from './types';

/** 曲目记录固定长度（0x174） */
export const TRACK_RECORD_SIZE = 372;
/** 根头长度 */
export const ROOT_HEADER_SIZE = 64;
/** hths 头长度（4s + I + I + Q） */
export const TRACK_HEADER_SIZE = 20;
/** 播放列表记录头长度 */
export const PLAYLIST_RECORD_HEADER = 44;
/** hphs 中偏移表之前的固定部分 */
export const PLAYLIST_HEADER_PREFIX = 68;

const MAGIC_ROOT = 'bdhs';
const MAGIC_TRACK_HEADER = 'hths';
const MAGIC_TRACK = 'rths';
const MAGIC_PLAYLIST_HEADER = 'hphs';
const MAGIC_PLAYLIST = 'lphs';

function magic(buf: Buffer, offset: number, expect: string): void {
  const got = buf.toString('latin1', offset, offset + 4);
  if (got !== expect) {
    throw new Error(
      `iTunesSD 结构异常：偏移 ${offset} 期望 '${expect}'，实际 '${got}'。` +
        `文件可能不是本工具支持的格式版本。`,
    );
  }
}

// ---------------------------------------------------------------- 解析

export function parseSd(data: Buffer): SdModel {
  if (data.length < ROOT_HEADER_SIZE) {
    throw new Error(`iTunesSD 太短（${data.length} 字节）`);
  }
  magic(data, 0, MAGIC_ROOT);

  const root: SdRoot = {
    version: data.readUInt32LE(0x04),
    totalLen: data.readUInt32LE(0x08),
    nTracks: data.readUInt32LE(0x0c),
    nPlaylists: data.readUInt32LE(0x10),
    unkQ: data.readBigUInt64LE(0x14),
    maxVolume: data[0x1c],
    voiceover: data[0x1d],
    unkH: data.readUInt16LE(0x1e),
    tracksWoPodcasts: data.readUInt32LE(0x20),
    trackHeaderOffset: data.readUInt32LE(0x24),
    playlistHeaderOffset: data.readUInt32LE(0x28),
    tail: Buffer.from(data.subarray(0x2c, 0x40)),
  };

  // --- 曲目区 ---
  const trkOff = root.trackHeaderOffset;
  magic(data, trkOff, MAGIC_TRACK_HEADER);
  // 注意：头长度字段（+4）是 (20 + n*4)，不是整个曲目块长度；
  // 而 +12 处的 8 字节 Q 是 unknown1，偏移表紧跟在 20 字节头之后。
  const thN = data.readUInt32LE(trkOff + 0x08);
  if (thN !== root.nTracks) {
    // 不致命，但值得暴露出来，便于诊断被外部工具改坏的文件
    console.warn(
      `[itunessd] hths 曲目数(${thN}) 与根头 nTracks(${root.nTracks}) 不一致`,
    );
  }
  const offsets: number[] = [];
  for (let i = 0; i < thN; i++) {
    const p = trkOff + TRACK_HEADER_SIZE + i * 4;
    if (p + 4 > data.length) {
      throw new Error('iTunesSD 曲目偏移表越界');
    }
    offsets.push(data.readUInt32LE(p));
  }

  const tracks: TrackRecord[] = offsets.map((o) => {
    if (o + TRACK_RECORD_SIZE > data.length) {
      throw new Error(`曲目记录越界：偏移 ${o}`);
    }
    return parseTrack(data.subarray(o, o + TRACK_RECORD_SIZE));
  });

  // --- 播放列表区 ---
  const plOff = root.playlistHeaderOffset;
  magic(data, plOff, MAGIC_PLAYLIST_HEADER);
  const plCount = data.readUInt32LE(plOff + 0x08);
  const headerSize = PLAYLIST_HEADER_PREFIX + plCount * 4;
  const playlistHeader = Buffer.from(data.subarray(plOff, plOff + headerSize));

  const playlists: PlaylistRecord[] = [];
  for (let k = 0; k < plCount; k++) {
    const slot = plOff + PLAYLIST_HEADER_PREFIX + k * 4;
    if (slot + 4 > data.length) break;
    const lo = data.readUInt32LE(slot);
    if (!lo) continue;
    playlists.push(parsePlaylist(data, lo));
  }

  return { root, tracks, playlistHeader, playlists };
}

export function parseTrack(r: Buffer): TrackRecord {
  magic(r, 0, MAGIC_TRACK);
  const filenameRaw = r.subarray(0x18, 0x118);
  const nul = filenameRaw.indexOf(0);
  return {
    headerLength: r.readUInt32LE(0x04),
    startMs: r.readUInt32LE(0x08),
    stopMs: r.readUInt32LE(0x0c),
    volumeGain: r.readUInt32LE(0x10),
    filetype: r.readUInt32LE(0x14),
    filename: filenameRaw
      .subarray(0, nul < 0 ? filenameRaw.length : nul)
      .toString('utf8'),
    bookmark: r.readUInt32LE(0x118),
    dontskip: r[0x11c],
    remember: r[0x11d],
    unintalbum: r[0x11e],
    unknown: r[0x11f],
    pregap: r.readUInt32LE(0x120),
    postgap: r.readUInt32LE(0x124),
    numsamples: r.readUInt32LE(0x128),
    unk12c: r.readUInt32LE(0x12c),
    audioBytes: r.readUInt32LE(0x130),
    unk134: r.readUInt32LE(0x134),
    albumid: r.readUInt32LE(0x138),
    trackNo: r.readUInt16LE(0x13c),
    disc: r.readUInt16LE(0x13e),
    unk140: r.readBigUInt64LE(0x140),
    dbid: Buffer.from(r.subarray(0x148, 0x150)),
    artistid: r.readUInt32LE(0x150),
    tail: Buffer.from(r.subarray(0x154, 0x174)),
  };
}

export function parsePlaylist(data: Buffer, off: number): PlaylistRecord {
  magic(data, off, MAGIC_PLAYLIST);
  const nSongs = data.readUInt32LE(off + 0x08);
  const members: number[] = [];
  for (let i = 0; i < nSongs; i++) {
    const p = off + PLAYLIST_RECORD_HEADER + i * 4;
    if (p + 4 > data.length) break;
    members.push(data.readUInt32LE(p));
  }
  return {
    header: Buffer.from(data.subarray(off, off + PLAYLIST_RECORD_HEADER)),
    totalLength: data.readUInt32LE(off + 0x04),
    nSongs,
    nNonaudio: data.readUInt32LE(off + 0x0c),
    dbid: Buffer.from(data.subarray(off + 0x10, off + 0x18)),
    listtype: data.readUInt32LE(off + 0x18),
    members,
  };
}

// ---------------------------------------------------------------- 序列化

export function buildTrack(t: TrackRecord): Buffer {
  const out = Buffer.alloc(TRACK_RECORD_SIZE);
  out.write(MAGIC_TRACK, 0, 'latin1');
  out.writeUInt32LE(t.headerLength || TRACK_RECORD_SIZE, 0x04);
  out.writeUInt32LE(t.startMs, 0x08);
  out.writeUInt32LE(t.stopMs, 0x0c);
  out.writeUInt32LE(t.volumeGain, 0x10);
  out.writeUInt32LE(t.filetype, 0x14);

  const name = Buffer.from(t.filename, 'utf8');
  if (name.length > 255) {
    throw new Error(`曲目路径过长（${name.length} > 255 字节）：${t.filename}`);
  }
  name.copy(out, 0x18);

  out.writeUInt32LE(t.bookmark, 0x118);
  out[0x11c] = t.dontskip;
  out[0x11d] = t.remember;
  out[0x11e] = t.unintalbum;
  out[0x11f] = t.unknown;
  out.writeUInt32LE(t.pregap, 0x120);
  out.writeUInt32LE(t.postgap, 0x124);
  out.writeUInt32LE(t.numsamples, 0x128);
  out.writeUInt32LE(t.unk12c, 0x12c);
  out.writeUInt32LE(t.audioBytes, 0x130);
  out.writeUInt32LE(t.unk134, 0x134);
  out.writeUInt32LE(t.albumid, 0x138);
  out.writeUInt16LE(t.trackNo, 0x13c);
  out.writeUInt16LE(t.disc, 0x13e);
  out.writeBigUInt64LE(t.unk140, 0x140);
  t.dbid.copy(out, 0x148, 0, 8);
  out.writeUInt32LE(t.artistid, 0x150);
  t.tail.copy(out, 0x154, 0, 32);
  return out;
}

/**
 * 序列化整个 iTunesSD。
 *
 * 头部模板整体沿用设备现有的值（版本标记、VoiceOver 开关、最大音量、模板尾部），
 * **不硬编码任何版本常量** —— 这样既适配本机的 0x02010001，也兼容其他代数。
 */
export function buildSd(model: SdModel): Buffer {
  const { tracks, root, playlists } = model;
  const n = tracks.length;

  const out = Buffer.alloc(ROOT_HEADER_SIZE);
  out.write(MAGIC_ROOT, 0, 'latin1');
  out.writeUInt32LE(root.version, 0x04);
  out.writeUInt32LE(ROOT_HEADER_SIZE, 0x08);
  out.writeUInt32LE(n, 0x0c);
  out.writeUInt32LE(root.nPlaylists, 0x10);
  out.writeBigUInt64LE(0n, 0x14);
  out[0x1c] = root.maxVolume & 0xff;
  out[0x1d] = root.voiceover ? 1 : 0; // VoiceOver 总开关，必须原样保留
  out.writeUInt16LE(0, 0x1e);
  out.writeUInt32LE(n, 0x20);
  root.tail.copy(out, 0x2c, 0, 20);

  // --- 曲目块 ---
  const thSize = TRACK_HEADER_SIZE + n * 4;
  const hunks: Buffer[] = [];
  const offsets = Buffer.alloc(n * 4);
  let cursor = ROOT_HEADER_SIZE + thSize;
  for (let i = 0; i < n; i++) {
    offsets.writeUInt32LE(cursor, i * 4);
    hunks.push(buildTrack(tracks[i]));
    cursor += TRACK_RECORD_SIZE;
  }
  const trk = Buffer.concat([
    headerChunk(MAGIC_TRACK_HEADER, thSize, n),
    offsets,
    ...hunks,
  ]);

  // --- 播放列表头 ---
  const ph = Buffer.from(model.playlistHeader);
  ph.writeUInt32LE(ph.length, 0x04);
  ph.writeUInt32LE(playlists.length, 0x08);
  const firstLphs = ROOT_HEADER_SIZE + trk.length + ph.length;
  for (let k = 0; k < playlists.length; k++) {
    const slot = PLAYLIST_HEADER_PREFIX + k * 4;
    if (slot + 4 > ph.length) break;
    ph.writeUInt32LE(k === 0 ? firstLphs : 0, slot);
  }

  out.writeUInt32LE(ROOT_HEADER_SIZE, 0x24);
  out.writeUInt32LE(ROOT_HEADER_SIZE + trk.length, 0x28);

  // --- 播放列表记录 ---
  const plParts: Buffer[] = [];
  for (const p of playlists) {
    const h = Buffer.from(p.header);
    h.writeUInt32LE(PLAYLIST_RECORD_HEADER + p.nSongs * 4, 0x04);
    h.writeUInt32LE(p.nSongs, 0x08);
    h.writeUInt32LE(p.nNonaudio, 0x0c);
    const mem = Buffer.alloc(p.nSongs * 4);
    p.members.forEach((m, i) => mem.writeUInt32LE(m, i * 4));
    plParts.push(h, mem);
  }

  return Buffer.concat([out, trk, ph, ...plParts]);
}

/** hths 头：magic + 偏移表长度 + 曲目数 + 8 字节 unknown1 */
function headerChunk(magic_: string, tableSize: number, count: number): Buffer {
  const b = Buffer.alloc(TRACK_HEADER_SIZE);
  b.write(magic_, 0, 'latin1');
  b.writeUInt32LE(tableSize, 0x04);
  b.writeUInt32LE(count, 0x08);
  b.writeBigUInt64LE(0n, 0x0c);
  return b;
}

// ---------------------------------------------------------------- 工具

/** 生成全新的曲目记录（用于新增文件）。存量曲目一律沿用设备原记录。 */
export function makeTrackRecord(opts: {
  filename: string;
  filetype: number;
  durationMs: number;
  sampleRate: number;
  audioBytes: number;
  dbid: Buffer;
  pregap: number;
  albumid: number;
  artistid: number;
  trackNo?: number;
  disc?: number;
  volumeGain?: number;
}): TrackRecord {
  return {
    headerLength: TRACK_RECORD_SIZE,
    startMs: 0,
    stopMs: Math.max(0, Math.round(opts.durationMs)),
    volumeGain: opts.volumeGain ?? 0,
    filetype: opts.filetype,
    filename: opts.filename,
    bookmark: 0,
    // 设备上 Apple 写入的 13 条记录该字节恒为 1，沿用以免参与随机播放行为异常
    dontskip: 1,
    remember: 0,
    unintalbum: 0,
    unknown: 0,
    pregap: opts.pregap,
    postgap: 0,
    numsamples: Math.round((opts.durationMs / 1000) * opts.sampleRate),
    unk12c: 0,
    audioBytes: opts.audioBytes,
    unk134: 0,
    albumid: opts.albumid,
    trackNo: opts.trackNo ?? 0,
    disc: opts.disc ?? 0,
    unk140: 0n,
    dbid: opts.dbid,
    artistid: opts.artistid,
    tail: Buffer.alloc(32),
  };
}

/** 用设备现有曲目重排成新的完整模型（曲目全部保留、顺序不变，重写 Master 列表）。 */
export function withTracks(
  template: SdModel,
  tracks: TrackRecord[],
  maxVolume = template.root.maxVolume,
): SdModel {
  const master = template.playlists[0];
  const header = master
    ? Buffer.from(master.header)
    : Buffer.alloc(PLAYLIST_RECORD_HEADER);
  if (!master) {
    header.write(MAGIC_PLAYLIST, 0, 'latin1');
  }
  const playlists: PlaylistRecord[] = [
    {
      header,
      totalLength: PLAYLIST_RECORD_HEADER + tracks.length * 4,
      nSongs: tracks.length,
      nNonaudio: tracks.length,
      dbid: Buffer.alloc(8),
      listtype: master ? master.listtype : 1,
      members: tracks.map((_, i) => i),
    },
  ];
  return {
    root: { ...template.root, nTracks: tracks.length, nPlaylists: 1, maxVolume },
    tracks,
    playlistHeader: Buffer.from(template.playlistHeader),
    playlists,
  };
}

export { FileType };

File: tauri-app/src-tauri/src/itunessd.rs
use crate::types::*;

pub const TRACK_RECORD_SIZE: usize = 372;
pub const ROOT_HEADER_SIZE: usize = 64;
pub const TRACK_HEADER_SIZE: usize = 20;
pub const PLAYLIST_RECORD_HEADER: usize = 44;
pub const PLAYLIST_HEADER_PREFIX: usize = 68;

const MAGIC_ROOT: &[u8; 4] = b"bdhs";
const MAGIC_TRACK_HEADER: &[u8; 4] = b"hths";
const MAGIC_TRACK: &[u8; 4] = b"rths";
const MAGIC_PLAYLIST_HEADER: &[u8; 4] = b"hphs";
const MAGIC_PLAYLIST: &[u8; 4] = b"lphs";

fn check_magic(data: &[u8], offset: usize, expected: &[u8; 4]) -> Result<(), String> {
    if offset + 4 > data.len() {
        return Err(format!("iTunesSD: offset {} out of bounds (len={})", offset, data.len()));
    }
    if &data[offset..offset+4] != expected {
        let got = String::from_utf8_lossy(&data[offset..offset+4]);
        let exp = String::from_utf8_lossy(expected);
        return Err(format!("iTunesSD: expected '{}' at offset {}, got '{}'", exp, offset, got));
    }
    Ok(())
}

fn rd_u16_le(data: &[u8], off: usize) -> u16 {
    u16::from_le_bytes([data[off], data[off+1]])
}
fn rd_u32_le(data: &[u8], off: usize) -> u32 {
    u32::from_le_bytes([data[off], data[off+1], data[off+2], data[off+3]])
}
fn rd_u64_le(data: &[u8], off: usize) -> u64 {
    u64::from_le_bytes([data[off], data[off+1], data[off+2], data[off+3],
                         data[off+4], data[off+5], data[off+6], data[off+7]])
}

fn read_utf16le_string(data: &[u8], off: usize, max_bytes: usize) -> String {
    let end = (off + max_bytes).min(data.len());
    let raw = &data[off..end];
    let u16s: Vec<u16> = raw.chunks_exact(2)
        .map(|c| u16::from_le_bytes([c[0], c[1]]))
        .take_while(|&c| c != 0)
        .collect();
    String::from_utf16_lossy(&u16s)
}

fn write_utf16le(buf: &mut [u8], off: usize, max_bytes: usize, s: &str) {
    let u16s: Vec<u16> = s.encode_utf16().collect();
    let max_chars = max_bytes / 2;
    let chars_to_write = u16s.len().min(max_chars.saturating_sub(1));
    for i in 0..chars_to_write {
        let bytes = u16s[i].to_le_bytes();
        buf[off + i * 2] = bytes[0];
        buf[off + i * 2 + 1] = bytes[1];
    }
}

pub fn parse_sd(data: &[u8]) -> Result<SdModel, String> {
    if data.len() < ROOT_HEADER_SIZE {
        return Err(format!("iTunesSD too short: {} bytes", data.len()));
    }
    check_magic(data, 0, MAGIC_ROOT)?;

    let mut tail = [0u8; 20];
    tail.copy_from_slice(&data[0x2c..0x40]);

    let root = SdRoot {
        version: rd_u32_le(data, 0x04),
        total_len: rd_u32_le(data, 0x08),
        n_tracks: rd_u32_le(data, 0x0c),
        n_playlists: rd_u32_le(data, 0x10),
        unk_q: rd_u64_le(data, 0x14),
        max_volume: data[0x1c],
        voiceover: data[0x1d],
        unk_h: rd_u16_le(data, 0x1e),
        tracks_wo_podcasts: rd_u32_le(data, 0x20),
        track_header_offset: rd_u32_le(data, 0x24),
        playlist_header_offset: rd_u32_le(data, 0x28),
        tail,
    };

    let trk_off = root.track_header_offset as usize;
    check_magic(data, trk_off, MAGIC_TRACK_HEADER)?;
    let th_n = rd_u32_le(data, trk_off + 0x08) as usize;

    let mut offsets = Vec::with_capacity(th_n);
    for i in 0..th_n {
        let p = trk_off + TRACK_HEADER_SIZE + i * 4;
        if p + 4 > data.len() {
            return Err("iTunesSD: track offset table out of bounds".to_string());
        }
        offsets.push(rd_u32_le(data, p) as usize);
    }

    let mut tracks = Vec::with_capacity(th_n);
    for &o in &offsets {
        if o + TRACK_RECORD_SIZE > data.len() {
            return Err(format!("iTunesSD: track record out of bounds at offset {}", o));
        }
        tracks.push(parse_track(&data[o..o + TRACK_RECORD_SIZE])?);
    }

    let pl_off = root.playlist_header_offset as usize;
    check_magic(data, pl_off, MAGIC_PLAYLIST_HEADER)?;
    let pl_count = rd_u32_le(data, pl_off + 0x08) as usize;
    let header_size = PLAYLIST_HEADER_PREFIX + pl_count * 4;
    let playlist_header = data[pl_off..pl_off + header_size].to_vec();

    let mut playlists = Vec::new();
    for k in 0..pl_count {
        let slot = pl_off + PLAYLIST_HEADER_PREFIX + k * 4;
        if slot + 4 > data.len() { break; }
        let lo = rd_u32_le(data, slot) as usize;
        if lo == 0 { continue; }
        playlists.push(parse_playlist(data, lo)?);
    }

    Ok(SdModel { root, tracks, playlist_header, playlists })
}

fn parse_track(r: &[u8]) -> Result<TrackRecord, String> {
    check_magic(r, 0, MAGIC_TRACK)?;

    let filename = read_utf16le_string(r, 0x18, 512);

    let mut dbid = [0u8; 8];
    dbid.copy_from_slice(&r[0x148..0x150]);

    let mut tail = [0u8; 32];
    tail.copy_from_slice(&r[0x154..0x174]);

    Ok(TrackRecord {
        header_length: rd_u32_le(r, 0x04),
        start_ms: rd_u32_le(r, 0x08),
        stop_ms: rd_u32_le(r, 0x0c),
        volume_gain: rd_u32_le(r, 0x10),
        filetype: rd_u32_le(r, 0x14),
        filename,
        bookmark: rd_u32_le(r, 0x118),
        dontskip: r[0x11c],
        remember: r[0x11d],
        unintalbum: r[0x11e],
        unknown_byte: r[0x11f],
        pregap: rd_u32_le(r, 0x120),
        postgap: rd_u32_le(r, 0x124),
        numsamples: rd_u32_le(r, 0x128),
        unk12c: rd_u32_le(r, 0x12c),
        audio_bytes: rd_u32_le(r, 0x130),
        unk134: rd_u32_le(r, 0x134),
        albumid: rd_u32_le(r, 0x138),
        track_no: rd_u16_le(r, 0x13c),
        disc: rd_u16_le(r, 0x13e),
        unk140: rd_u64_le(r, 0x140),
        dbid,
        artistid: rd_u32_le(r, 0x150),
        tail,
    })
}

fn parse_playlist(data: &[u8], off: usize) -> Result<PlaylistRecord, String> {
    if off + PLAYLIST_RECORD_HEADER > data.len() {
        return Err("iTunesSD: playlist record out of bounds".to_string());
    }
    check_magic(data, off, MAGIC_PLAYLIST)?;

    let total_length = rd_u32_le(data, off + 0x04);
    let n_songs = rd_u32_le(data, off + 0x08);
    let n_nonaudio = rd_u32_le(data, off + 0x0c);
    let mut dbid = [0u8; 8];
    dbid.copy_from_slice(&data[off + 0x10..off + 0x18]);
    let listtype = rd_u32_le(data, off + 0x18);
    let header_end = off + PLAYLIST_RECORD_HEADER;
    let header = data[off..header_end].to_vec();

    let mut members = Vec::with_capacity(n_songs as usize);
    for i in 0..n_songs as usize {
        let p = header_end + i * 4;
        if p + 4 > data.len() { break; }
        members.push(rd_u32_le(data, p));
    }

    Ok(PlaylistRecord { header, total_length, n_songs, n_nonaudio, dbid, listtype, members })
}

fn build_track_header(total_size: u32, n_tracks: u32) -> Vec<u8> {
    let mut h = vec![0u8; TRACK_HEADER_SIZE];
    h[0..4].copy_from_slice(MAGIC_TRACK_HEADER);
    h[4..8].copy_from_slice(&total_size.to_le_bytes());
    h[8..12].copy_from_slice(&n_tracks.to_le_bytes());
    h
}

fn write_track(t: &TrackRecord) -> Vec<u8> {
    let mut buf = vec![0u8; TRACK_RECORD_SIZE];
    buf[0..4].copy_from_slice(MAGIC_TRACK);
    buf[0x04..0x08].copy_from_slice(&t.header_length.to_le_bytes());
    buf[0x08..0x0c].copy_from_slice(&t.start_ms.to_le_bytes());
    buf[0x0c..0x10].copy_from_slice(&t.stop_ms.to_le_bytes());
    buf[0x10..0x14].copy_from_slice(&t.volume_gain.to_le_bytes());
    buf[0x14..0x18].copy_from_slice(&t.filetype.to_le_bytes());
    write_utf16le(&mut buf, 0x18, 512, &t.filename);
    buf[0x118..0x11c].copy_from_slice(&t.bookmark.to_le_bytes());
    buf[0x11c] = t.dontskip;
    buf[0x11d] = t.remember;
    buf[0x11e] = t.unintalbum;
    buf[0x11f] = t.unknown_byte;
    buf[0x120..0x124].copy_from_slice(&t.pregap.to_le_bytes());
    buf[0x124..0x128].copy_from_slice(&t.postgap.to_le_bytes());
    buf[0x128..0x12c].copy_from_slice(&t.numsamples.to_le_bytes());
    buf[0x12c..0x130].copy_from_slice(&t.unk12c.to_le_bytes());
    buf[0x130..0x134].copy_from_slice(&t.audio_bytes.to_le_bytes());
    buf[0x134..0x138].copy_from_slice(&t.unk134.to_le_bytes());
    buf[0x138..0x13c].copy_from_slice(&t.albumid.to_le_bytes());
    buf[0x13c..0x13e].copy_from_slice(&t.track_no.to_le_bytes());
    buf[0x13e..0x140].copy_from_slice(&t.disc.to_le_bytes());
    buf[0x140..0x148].copy_from_slice(&t.unk140.to_le_bytes());
    buf[0x148..0x150].copy_from_slice(&t.dbid);
    buf[0x150..0x154].copy_from_slice(&t.artistid.to_le_bytes());
    buf[0x154..0x174].copy_from_slice(&t.tail);
    buf
}

pub fn build_sd(model: &SdModel) -> Vec<u8> {
    let n = model.tracks.len();
    let mut out = vec![0u8; ROOT_HEADER_SIZE];
    out[0..4].copy_from_slice(MAGIC_ROOT);
    out[0x04..0x08].copy_from_slice(&model.root.version.to_le_bytes());
    out[0x0c..0x10].copy_from_slice(&(n as u32).to_le_bytes());
    out[0x10..0x14].copy_from_slice(&model.root.n_playlists.to_le_bytes());
    out[0x14..0x1c].copy_from_slice(&model.root.unk_q.to_le_bytes());
    out[0x1c] = model.root.max_volume;
    out[0x1d] = if model.root.voiceover != 0 { 1 } else { 0 };
    out[0x1e..0x20].copy_from_slice(&0u16.to_le_bytes());
    out[0x20..0x24].copy_from_slice(&(n as u32).to_le_bytes());
    out[0x2c..0x40].copy_from_slice(&model.root.tail);

    let th_size = TRACK_HEADER_SIZE + n * 4;
    let mut offsets_buf = vec![0u8; n * 4];
    let mut track_hunks = Vec::new();
    let mut cursor = ROOT_HEADER_SIZE + th_size;
    for i in 0..n {
        offsets_buf[i*4..i*4+4].copy_from_slice(&(cursor as u32).to_le_bytes());
        track_hunks.push(write_track(&model.tracks[i]));
        cursor += TRACK_RECORD_SIZE;
    }

    let th_header = build_track_header(th_size as u32, n as u32);
    let mut trk = Vec::new();
    trk.extend_from_slice(&th_header);
    trk.extend_from_slice(&offsets_buf);
    for h in &track_hunks {
        trk.extend_from_slice(h);
    }

    let mut ph = model.playlist_header.clone();
    let ph_len = ph.len();
    if ph_len >= 12 {
        ph[0x04..0x08].copy_from_slice(&(ph_len as u32).to_le_bytes());
        ph[0x08..0x0c].copy_from_slice(&(model.playlists.len() as u32).to_le_bytes());
    }
    let first_lphs = ROOT_HEADER_SIZE + trk.len() + ph_len;
    let pl_count_in_header = model.playlists.len().min((ph_len.saturating_sub(PLAYLIST_HEADER_PREFIX)) / 4);
    for k in 0..pl_count_in_header {
        let slot = PLAYLIST_HEADER_PREFIX + k * 4;
        if slot + 4 > ph_len { break; }
        let val = if k == 0 { first_lphs as u32 } else { 0u32 };
        ph[slot..slot+4].copy_from_slice(&val.to_le_bytes());
    }

    out[0x24..0x28].copy_from_slice(&(ROOT_HEADER_SIZE as u32).to_le_bytes());
    out[0x28..0x2c].copy_from_slice(&((ROOT_HEADER_SIZE + trk.len()) as u32).to_le_bytes());

    let mut pl_parts = Vec::new();
    for p in &model.playlists {
        let mut h = p.header.clone();
        if h.len() >= PLAYLIST_RECORD_HEADER {
            let rec_len = (PLAYLIST_RECORD_HEADER + p.n_songs as usize * 4) as u32;
            h[0x04..0x08].copy_from_slice(&rec_len.to_le_bytes());
            h[0x08..0x0c].copy_from_slice(&p.n_songs.to_le_bytes());
            h[0x0c..0x10].copy_from_slice(&p.n_nonaudio.to_le_bytes());
        }
        let mut mem = Vec::with_capacity(p.members.len() * 4);
        for &m in &p.members {
            mem.extend_from_slice(&m.to_le_bytes());
        }
        pl_parts.push(h);
        pl_parts.push(mem);
    }

    let mut result = Vec::new();
    result.extend_from_slice(&out);
    result.extend_from_slice(&trk);
    result.extend_from_slice(&ph);
    for part in &pl_parts {
        result.extend_from_slice(part);
    }

    let total_len = result.len() as u32;
    result[0x08..0x0c].copy_from_slice(&total_len.to_le_bytes());
    result
}

pub fn make_track_record(
    filename: &str,
    filetype: u32,
    duration_ms: u32,
    sample_rate: u32,
    audio_bytes: u32,
    dbid: [u8; 8],
    pregap: u32,
    albumid: u32,
    artistid: u32,
    track_no: u16,
    disc: u16,
    volume_gain: u32,
) -> TrackRecord {
    let duration_secs = duration_ms as f64 / 1000.0;
    TrackRecord {
        header_length: TRACK_RECORD_SIZE as u32,
        start_ms: 0,
        stop_ms: duration_ms,
        volume_gain,
        filetype,
        filename: filename.to_string(),
        bookmark: 0,
        dontskip: 1,
        remember: 0,
        unintalbum: 0,
        unknown_byte: 0,
        pregap,
        postgap: 0,
        numsamples: (duration_secs * sample_rate as f64).round() as u32,
        unk12c: 0,
        audio_bytes,
        unk134: 0,
        albumid,
        track_no,
        disc,
        unk140: 0,
        dbid,
        artistid,
        tail: [0u8; 32],
    }
}

pub fn with_tracks(template: &SdModel, tracks: Vec<TrackRecord>, max_volume: Option<u8>) -> SdModel {
    let mv = max_volume.unwrap_or(template.root.max_volume);
    let master = template.playlists.first();
    let header = if let Some(m) = master {
        m.header.clone()
    } else {
        let mut h = vec![0u8; PLAYLIST_RECORD_HEADER];
        h[0..4].copy_from_slice(MAGIC_PLAYLIST);
        h
    };
    let listtype = master.map(|m| m.listtype).unwrap_or(1);
    let n = tracks.len() as u32;
    let members: Vec<u32> = (0..n).collect();

    let playlists = vec![PlaylistRecord {
        header,
        total_length: (PLAYLIST_RECORD_HEADER + tracks.len() * 4) as u32,
        n_songs: n,
        n_nonaudio: n,
        dbid: [0u8; 8],
        listtype,
        members,
    }];

    SdModel {
        root: SdRoot {
            n_tracks: n,
            n_playlists: 1,
            max_volume: mv,
            ..template.root.clone()
        },
        tracks,
        playlist_header: template.playlist_header.clone(),
        playlists,
    }
}

pub fn round_trips(sd_path: &std::path::Path) -> bool {
    let raw = match std::fs::read(sd_path) {
        Ok(d) => d,
        Err(_) => return false,
    };
    match parse_sd(&raw) {
        Ok(model) => build_sd(&model) == raw,
        Err(_) => false,
    }
}
